#167190531 Implement User Reset Password Via Email

ayodejiAA opened the pull request:
#### What does this PR do?
Add routes for users to request reset password link and use the link to reset password

#### Description of Task proposed in this pull request?
- Create a route to request and send password reset link to the user email address
- Create another route to reset password with the link sent
- Add validations for user inputs 
- Add model, unit and integration tests
- Document this feature with swagger

#### How should this be manually tested (Quality Assurance)?
- Register as a new user with   -  **POST** ``` /users/create/```
- ```POST``` request to ```/users/resetpassword``` and add your email address to the request body. ```Schema``` - ```"email" : "youreamiladdress```. A link will be sent to the email address your provided. 
- Copy the link and post to postman. Required fields for new password is 
```password: newPassword``` 
```confirmPassword:  newPassword```


#### What are the relevant pivotal tracker stories?
- [#167190531](https://www.pivotaltracker.com/story/show/167190531)

#### Any background context you want to add (Operations Impact)?
- Make sure all required packages are installed and environment variables are provided as well. 

#### What I have learned working on this feature: 
- I learnt a few things working with Sequelize models and a lot from unit testing

#### Screenshots: 
<img width="1030" alt="Screenshot 2019-08-09 at 10 34 40 AM" src="https://user-images.githubusercontent.com/42908597/62770366-cdaa6a80-ba92-11e9-9f42-b6ef911f8567.png">
<img width="1039" alt="Screenshot 2019-08-09 at 10 40 33 AM" src="https://user-images.githubusercontent.com/42908597/62770398-dbf88680-ba92-11e9-85cb-aa0d7f481117.png">
<img width="1199" alt="Screenshot 2019-08-09 at 10 39 51 AM" src="https://user-images.githubusercontent.com/42908597/62770412-e450c180-ba92-11e9-95c7-a8d18518da7d.png">

chialuka commented on server/docs/authors-haven-api.yml:
```diff
@@ -266,5 +352,17 @@ components:
+          type: string
 
 
+    
```
Please add a new line here

ayodejiAA replied:
Done

chialuka commented on server/middlewares/userValidation.js:
```diff
@@ -1,21 +1,37 @@
-  const options = {
-    abortEarly: false
-  };
+const validate = (schema) => {
```
Could you consider passing the express middleware functions to this validate function as well so that we have one validation function and not two?

ayodejiAA replied:
I believe we will still have two functions accepting parameters, but the es6 syntax looks good.

dinobi commented on server/controllers/Users.js:
```diff
@@ -86,6 +89,72 @@ class Users {
+      }
+      const { id } = user;
+      const token = generateToken({ id }, '1h');
+      await ResetPasswordToken.create({ token, userId: id });
```
Please change table to`ResetPassword`

dinobi commented on server/controllers/Users.js:
```diff
@@ -86,6 +89,72 @@ class Users {
+      const { id } = user;
+      const token = generateToken({ id }, '1h');
+      await ResetPasswordToken.create({ token, userId: id });
+      await Session.deactivateAllById(id);
```
change this to `Session.revokeAll`

dinobi commented on server/controllers/Users.js:
```diff
@@ -86,6 +89,72 @@ class Users {
+      const token = generateToken({ id }, '1h');
+      await ResetPasswordToken.create({ token, userId: id });
+      await Session.deactivateAllById(id);
+      requestResetPasswordMail({ ...user, token });
```
change this is to `sendResetPasswordEmail`

ayodejiAA committed:
feature-[167190531]: Implement User Reset Password Via Email

- create route to request and send password reset link to user email
- create route to reset password
- add valiadations for user inputs
- add model,unit and integration tests
- document feature with swagger

[Delivers #167190531]

chialuka commented on server/controllers/Users.js:
```diff
@@ -111,6 +114,37 @@ class Users {
+   * @memberof Users
+   * @returns {JSON} JSON object
+   */
+  static async requestPasswordResetLink(req, res) {
```
The name of the function doesn't match the name on the JSDoc. Please fix

ayodejiAA replied:
Done.

Adekoreday commented on server/controllers/Users.js:
```diff
@@ -111,6 +114,37 @@ class Users {
+      const token = generateToken({ id }, '1h');
+      await ResetPassword.create({ token, userId: id });
+      await Session.revokeAll(id);
+      sendResetPasswordEmail({ ...user, token });
```
you need to await the sendResetPasswordEmail it's an async operation

JuwonAbiola reviewed (dismissed):
Looks good to me

JuwonAbiola reviewed (approved):
Looks good to me
